### ISSUE TITLE
Where is the documentation for the REST api?

### ISSUE THREAD

**Issue by Trusted97** [devpi/devpi#1042]
Hello there!
I would like to study the REST api but online i didn't find anything could you help me?

**Comment by fschulze** [comment]
There is no documentation for it yet. Do you have a specific goal, or is it out of curiosity?

**Comment by Trusted97** [comment]
I'm looking for a private Python registry and i'm curious to see what devpi is offering, REST api for retrieve all packages or single update would be great for me

**Comment by fschulze** [comment]
Getting info is pretty straight forward. Fetch json from root, /user, /user/index, /user/index/package and /user/index/package/version. For writing I recommend looking into ``devpi-tokens`` for easier authentication than the default. Then you can use the same API as for PyPI with a POST on /user/index (for which I can't find the documentation right now) or use twine etc to upload files. The only remaining API is for creating and modifying users and indexes, but that is a bit much to write at the moment.

**Comment by brian7704** [comment]
Is there an API endpoint that lists all of an index's packages with their metadata?

**Comment by fschulze** [comment]
@brian7704 No. It would also be pretty bad from a performance point of view when there are more than a couple of packages. This is already a problem for the index view of devpi-web in some cases. There are some plans to improve on that, but that won't allow good performance to include *all* metadata.
